src/ray/gcs/gcs_server/gcs_task_manager.h

```diff
@@ -128,15 +150,23 @@ class GcsTaskManager : public rpc::TaskInfoHandler {
     /// Max number of task events allowed in the storage.
     const size_t max_num_task_events_ = 0;
 
-    /// Current task events stored.
-    std::vector<rpc::TaskEvents> task_events_;
-
     /// A iterator into task_events_ that determines which element to be overwritten.
     size_t next_idx_to_overwrite_ = 0;
 
-    /// Index from task attempt to the index of the corresponding task event.
+    /// Current task events stored.
+    std::vector<rpc::TaskEvents> task_events_;
+
+    /// Index from task attempt to the corresponding task attempt in the buffer
+    /// `task_events_`.
     absl::flat_hash_map<TaskAttempt, size_t> task_attempt_index_;
 
+    /// Secondary index from task id to task attempts.
+    absl::flat_hash_map<TaskID, absl::flat_hash_set<TaskAttempt>>
```

Comment:
Actually I am curious how difficult it is to use https://www.boost.org/doc/libs/1_61_0/libs/multi_index/doc/index.html?

Reply:
(not a blocker for the PR)

Reply:
or is there any way to build a clean interface to combine these  4 different indexes?

Reply:
Let me look into this - yeah, or potentially I could abstract these indexes into an abstraction where that `Index` will allow get by different things.